### PR TITLE
ascii: only support checking String for pure ASCIIness

### PR DESCRIPTION
**Pull request by StefanKarpinski**
Part of #16107.

### CODE REVIEW

File: base/strings/util.jl [hunk 1]
```diff
@@ -243,3 +243,13 @@ function bytes2hex(a::AbstractArray{UInt8})
     end
     return String(b)
 end
+
+# check for pure ASCII-ness
+
+function ascii(s::String)
+    for b in s.data
+        b < 0x80 || throw(ArgumentError("invalid ASCII sequence"))
```
Comment (tkelman) [inline base/strings/util.jl]: should we show the string?

**Comment by tkelman** [comment]
should add a test that it throws on non-ascii input, and docstrings need updating

File: base/strings/util.jl (hunk 1 above)
Comment (nalimilan) [inline base/strings/util.jl]: +1, show the position where the invalid char was found and print it.

**Comment by StefanKarpinski** [comment]
What's up with the test failure on AppVeyor? It looks like it succeeded and then hung.

File: base/docs/helpdb/Base.jl [hunk 2]
```diff
@@ -8719,29 +8719,14 @@ alias the input `x` to modify it in-place.
 filt!
 
 """
-    ascii(::Array{UInt8,1})
+    ascii(s::AbstractString)
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: make corresponding change in rst too

File: base/deprecated.jl [hunk 3]
```diff
@@ -1155,6 +1155,10 @@ end
 @deprecate_binding UTF8String String
 @deprecate_binding ByteString String
 
+@deprecate ascii(p::Ptr{UInt8}, len::Integer) ascii(bytestring(p, len))
```
Comment (stevengj) [inline base/deprecated.jl]: Shouldn't this be deprecated to `let s=bytestring(p,len); !isascii(s) && error("expecting ASCII string"); s; end`?

File: base/docs/helpdb/Base.jl (hunk 2 above)
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: would be good to migrate the remaining docstring inline

**Comment by StefanKarpinski** [comment]
Any thoughts on whether I should merge this now and do additional work in a new PR or just add commits here? By additional work, I mean more parts of the `replace utf8, bytestring, etc. with String` bullet in #16107.

File: base/docs/helpdb/Base.jl (hunk 2 above)
Comment (StefanKarpinski) [inline base/docs/helpdb/Base.jl]: Doesn't that get automatically done when the RST gets regenerated?

Comment (yuyichao) [inline base/docs/helpdb/Base.jl]: This is what the script uses to find the correct place to insert the docstring so changing the signature will make the doc generation fail.

Comment (StefanKarpinski) [inline base/docs/helpdb/Base.jl]: The signatures now match...

Comment (tkelman) [inline base/docs/helpdb/Base.jl]: By inline I mean moving the markdown docstring to the source, so we can gradually get rid of this Base.jl file.

File: base/strings/util.jl [hunk 4]
```diff
@@ -243,3 +243,13 @@ function bytes2hex(a::AbstractArray{UInt8})
     end
     return String(b)
 end
+
+# check for pure ASCII-ness
+
+function ascii(s::String)
+    for (i, b) in enumerate(s.data)
+        b < 0x80 || throw(ArgumentError("invalid ASCII at index $i in $(repr(s))"))
```
Comment (tkelman) [inline base/strings/util.jl]: do we have or want a more specific exception type than ArgumentError for this? I guess we have UnicodeError but are you planning on removing that?

**Comment by tkelman** [comment]
I like the idea of keeping them small and separate if possible. Easier to review in smaller pieces.

File: base/docs/helpdb/Base.jl (hunk 2 above)
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: You also should have run genstdlib and committed that at the same time, otherwise unrelated doc changes are ending up in multiple other PR's.
